feat: Dev Portal SSO one pager (#7633)
Kong/docs.konghq.com: app/konnect/dev-portal/access-and-approval/sso.md

---
title: Configure SSO for Dev Portal
content_type: how-to
---

You can configure single sign-on (SSO) for {{site.konnect_short_name}} Dev Portal with OIDC. This allows developers to log in to Dev Portals by using their SSO credentials. 

> <!--Do developers get auto approved in Dev Portal then?-->

Reviewer: Developers do get auto-approved by Konnect when using SSO. This is because we outsource the approval process to the IDP instance when using SSO. Meaning it is possible to restrict who can sign up via the IDP rather than through Konnect.

> <!-- would we recommend SSO instead of adding developers and teams from IdPs if we don't want to map team? Or would we maybe recommend both (one to create/map devs and the other to let them log in to Dev Portal)-->

Reviewer: SSO and IDP team mappings must come from the same IDP instance

> <!-- should we add a note that this SSO is different than the one for Konnect? As in, if they want to use SSO for logging in to Konnect, they still need to configure that separately?-->

Reviewer: 👍 That could be a good not with a link to the configuring in Konnect. Also could be good to note that each Dev Portal has a separate SSO configuration. They can use the same IDP for multiple portals or different IDPs per portal.

## Prerequisites

> * Ensure that any users that need to use the Dev Portal SSO have been added to the **User Management > Users** section of your Auth0 tenant

Reviewer: ```suggestion
* Ensure that any user emails that need to use the Dev Portal SSO have been added to as users in your IDP instance
```

> * An application for {{site.konnect_short_name}} configured in your IdP:

Reviewer: Note that other IDPs are able to be used if they follow the OIDC standard for SSO. Those listed are the most common

1. Navigate to the **Applications > Applications** page to and click "Create Application" create a new application.

    * Name the application, making it easy to reference as the app that manages your Konnect Dev Portal SSO integration.
    * Select **Machine To Machine** as the **Application Type**.
    * Select **Auth0 Management API** as the Authorized API.
    * Authorize the application for at least one Permission to continue, e.g. `read:client_grants`. Note: this permission can be revoked later if desired, but is required to create the application.

1. Submit the form to create the application.

## Configure the Auth0 application

1. Locate your portal's callback URL in the Konnect Identity Provider Configuration Settings page. It will have the form `https://<your_portal_domain>/login`
1. Navigate to the **Settings** tab and enter the value of your portal's callback URL into each of the following fields:

    * **Application Login URI**
    * **Allowed Callback URLs**

1. Scroll down to the **Advanced Settings** and open the **Grant Types** tab.

    * Check the box for **Authorization Code** and click **Save Changes**

## Configure Login Action in Auth0

   {:.note}

>    > **Important:** This section is required due to the Auth0 API implementation not being inline with the OIDC standard for the value of the `updated_at` token claim.

Reviewer: Looks like this notice got combined with the next list item since there is no newline between
```suggestion
   > **Important:** This section is required due to the Auth0 API implementation not being inline with the OIDC standard for the value of the `updated_at` token claim.
   
```

    * **Grant Types**: Authorization Code

1. Create a login [action in Auth0](https://auth0.com/docs/customize/actions/write-your-first-action#create-an-action) and enter the following in the Actions Code Editor:
    
    ```js
    exports.onExecutePostLogin = async (event, api) => {
    if (event.authorization) {
        // This transforms the ISO 8601 Timestamp string into the seconds integer representation that is expected for the OIDC standard,
        // allowing the Konnect SSO validation to accept the format of the `updated_at` property when parsing the token claim.
        api.idToken.setCustomClaim('updated_at', Math.floor(new Date(event.user.updated_at).getTime()/1000))
    }
    };
    ```

   {:.note}
   > **Important:** This section is required because the Auth0 API implementation isn't inline with the OIDC standard for the `updated_at` token claim value.

1. Deploy the action by dragging it from the Start to Complete step in the [Login Flow](https://auth0.com/docs/customize/actions/write-your-first-action#deploy-the-action).
{% endnavtab %}
{% endnavtabs %}

## Configure SSO in {{site.konnect_short_name}}

> From the [{{site.konnect_short_name}} portal identity page](https://cloud.konghq.com/portal/portal-settings#identity), click **Configure provider** for **OIDC**, and enter the values from your IdP application.

Reviewer 2: This link takes me to the portal settings page, then I have to click identity.

> From the [{{site.konnect_short_name}} portal identity page](https://cloud.konghq.com/portal/portal-settings#identity), click **Configure provider** for **OIDC**, and enter the values from your IdP application.

Author: ```suggestion
From the [{{site.konnect_short_name}} Dev Portal settings](https://cloud.konghq.com/portal/portal-settings), click the **Identity** tab and then click **Configure provider** for **OIDC**. Enter the values from your IdP application.
```

* If you plan on using [team mappings from an IdP](/konnect/dev-portal/access-and-approval/add-teams), they must be from the same IdP instance as your SSO.
* If you have multiple Dev Portals, keep in mind that each Dev Portal has a separate SSO configuration. You can use the same IdP for multiple Dev Portals or different IdPs per Dev Portal.
* Dev Portal SSO is different than the [SSO for {{site.konnect_short_name}}](/konnect/org-management/oidc-idp). If you want to use SSO to log in to {{site.konnect_short_name}}, you must configure that separately. 

## Prerequisites

* Ensure that any users that need to use the Dev Portal SSO have been added to your IdP tenant
* An application for {{site.konnect_short_name}} configured in your IdP. 
    * [Okta](https://help.okta.com/en-us/content/topics/apps/apps_app_integration_wizard.htm)
    * [Azure AD](https://learn.microsoft.com/graph/toolkit/get-started/add-aad-app-registration)
    * [Auth0](https://auth0.com/docs/get-started/auth0-overview/create-applications)

## Configure an application and group claims in your IdP
{% navtabs %}

> {% navtab Azure %}

Reviewer 2: The firs three steps of the Azure instructions feel confusing because of the links. 
If you want to take this approach, i think it might be better to combine the first three steps into one.

Reviewer 2: Like you did with the Okta tab.

> 1. In [Azure](https://portal.azure.com/), [create an application](https://learn.microsoft.com/en-us/entra/identity-platform/quickstart-register-app?tabs=certificate) for {{site.konnect_short_name}}. 

Reviewer 2: ```suggestion
1. In [Azure](https://portal.azure.com/), [register an application](https://learn.microsoft.com/en-us/entra/identity-platform/quickstart-register-app?tabs=certificate) for {{site.konnect_short_name}}. 
```
It looks like they use the word register in the official docs.

> 1. Enter the Dev Portal [Redirect URI](/konnect/dev-portal/access/) for the **Redirect URI**. 

Reviewer 2: Redirect URI here just links to: https://deploy-preview-7633--kongdocs.netlify.app/konnect/dev-portal/

> 1. [Create a client secret](https://learn.microsoft.com/en-us/entra/identity-platform/quickstart-register-app?tabs=client-secret) and save the secret value to configure {{site.konnect_short_name}}.

Reviewer 2: ```suggestion
1. [Create a client secret](https://learn.microsoft.com/en-us/entra/identity-platform/quickstart-register-app?tabs=client-secret) and save the secret value to configure {{site.konnect_short_name}}.
```
This doesn't link me to the create a client secret tab, and so instead it just opens the same link from the first step. That feels confusing to me.

1. [Use the OIDC well-known discovery endpoint](https://learn.microsoft.com/en-us/azure/active-directory-b2c/secure-api-management?tabs=app-reg-ga#get-a-token-issuer-endpoint) to find and save the `issuer` value. 
    The `issuer` value will be used as the provider URL when configuring SSO in {{site.konnect_short_name}}.

1. [Add a group claim](https://learn.microsoft.com/en-us/entra/identity-platform/optional-claims?tabs=appui#configure-groups-optional-claims). Enable all the group types settings and select **Group ID** for each setting in the token properties type.

1. [Configure an optional claim](https://learn.microsoft.com/en-us/entra/identity-platform/optional-claims?tabs=appui#configure-optional-claims-in-your-application) with **ID** as the token type and **email** as the claim.

{% endnavtab %}

> {% navtab Okta %}

Reviewer 2: This section works a lot better, I think the microsoft docs just are too messy.

    Field | Value
    ---|---
    Name | `groups`
    Include in token type | ID token, Always
    Value type | Groups
    Filter | Select **Matches regex** from the drop-down, then enter `.*` in the field.
    Include in | Choose **The following scopes** and select `openid`, `email`, and `profile`. 

    This claim tells Okta to reference a subset of Okta groups.
    In this case, the wildcard (`.*`) value tells Okta to make all groups
    available for team mapping.

    {:.important}
    > If the authorization server is pulling in additional groups from
    third-party applications (for example, Google groups), the `groups` claim
    cannot find them. An Okta administrator needs to duplicate those groups and
    re-create them directly in Okta. They can do this by exporting the group in
    question in CSV format, then importing the CSV file to populate the new group.

> 1. [Add users to the application](https://help.okta.com/en-us/content/topics/users-groups-profiles/usgp-assign-apps.htm).

Reviewer 2: ```suggestion
1. [Add users to the Okta application](https://help.okta.com/en-us/content/topics/users-groups-profiles/usgp-assign-apps.htm).
```

Reviewer 2: Since we tab between two properties that both contain a user value we should be specific.

> 1. [Test ID token claims](https://developer.okta.com/docs/guides/customize-authz-server/main/#create-claims) and find groups for mapping. Configure the following claims settings:

Reviewer 2: Is this the right link?

{% endnavtab %}
{% navtab Auth0 %}
Kong offers OIDC support to allow Single-Sign-on for {{site.konnect_short_name}} and the Dev Portal. This guide shows you how to configure Auth0 for Dev Portal SSO.

1. [Create an application in Auth0](https://auth0.com/docs/get-started/auth0-overview/create-applications/machine-to-machine-apps) and configure the following settings:
    * **Application Type**: Machine to Machine Applications
    * **Authorized API**: Auth0 Management API
    * **Permissions**: Authorize the application for at least one permission, for example: `read:client_grants`. 
    This permission can be revoked later if desired, but one permission is required to create the application.

1. [Configure the Auth0 application settings](https://auth0.com/docs/get-started/applications/application-settings):
    * **Application Login URI**: `https://cloud.konghq.com/login`
    * **Allowed Callback URLs**: `https://cloud.konghq.com/login`
    * **Grant Types**: Authorization Code

> 1. [Create a login action in Auth0](https://auth0.com/docs/customize/actions/write-your-first-action#create-an-action) and enter the following in the Actions Code Editor:

Reviewer 2: ```suggestion
1. Create a login [action in Auth0](https://auth0.com/docs/customize/actions/write-your-first-action#create-an-action) and enter the following in the Actions Code Editor:
```

    
    ```js
    exports.onExecutePostLogin = async (event, api) => {
    if (event.authorization) {
        // This transforms the ISO 8601 Timestamp string into the seconds integer representation that is expected for the OIDC standard,
        // allowing the Konnect SSO validation to accept the format of the `updated_at` property when parsing the token claim.
        api.idToken.setCustomClaim('updated_at', Math.floor(new Date(event.user.updated_at).getTime()/1000))
    }
    };
    ```

   {:.note}
   > **Important:** This section is required because the Auth0 API implementation isn't inline with the OIDC standard for the `updated_at` token claim value.

> 1. Deploy the action by dragging it from the Start to Complete step in the [Login Flow](https://auth0.com/docs/customize/actions/flows-and-triggers/login-flow).

Reviewer 2: Should it be this link?
https://auth0.com/docs/customize/actions/write-your-first-action#deploy-the-action
